File: languages/compile-mo.py
#!/usr/bin/env python3
"""
Simple PO to MO compiler
"""

import sys
import struct

def compile_po_to_mo(po_file, mo_file=None):
    if mo_file is None:
        mo_file = po_file.replace('.po', '.mo')
    
    translations = {}
    current_msgid = ''
    current_msgstr = ''
    in_msgid = False
    in_msgstr = False
    
    with open(po_file, 'r', encoding='utf-8') as f:
        for line in f:
            line = line.strip()
            
            if not line or line.startswith('#'):
                continue
            
            if line.startswith('msgid '):
                if in_msgstr:
                    translations[current_msgid] = current_msgstr
                current_msgid = line[7:-1]  # Remove "msgid " and quotes
                current_msgstr = ''
                in_msgid = True
                in_msgstr = False
            elif line.startswith('msgstr '):
                current_msgstr = line[8:-1]  # Remove "msgstr " and quotes
                in_msgid = False
                in_msgstr = True
            elif in_msgid and line.startswith('"'):
                current_msgid += line[1:-1]
            elif in_msgstr and line.startswith('"'):
                current_msgstr += line[1:-1]
    
    if in_msgstr:
        translations[current_msgid] = current_msgstr
    
    # Create MO file
    num_entries = len(translations)
    
    # MO file header
    mo_data = struct.pack('<IIII', 
                          0x950412de,  # Magic number
                          0,           # Version
                          num_entries, # Number of entries
                          28)          # Offset of original strings table
    
    # Prepare string data
    original_strings = b''
    translation_strings = b''
    original_offsets = []
    translation_offsets = []
    
    for original, translation in translations.items():
        original_bytes = original.encode('utf-8')
        translation_bytes = translation.encode('utf-8')
        
        original_offsets.append((len(original_bytes), len(original_strings)))
        original_strings += original_bytes + b'\0'
        
        translation_offsets.append((len(translation_bytes), len(translation_strings)))
        translation_strings += translation_bytes + b'\0'
    
    # Write original strings table
    for length, offset in original_offsets:
        mo_data += struct.pack('<II', length, offset)
    
    # Write translation strings table
    for length, offset in translation_offsets:
        mo_data += struct.pack('<II', length, offset)
    
    # Write string data
    mo_data += original_strings
    mo_data += translation_strings
    
    with open(mo_file, 'wb') as f:
        f.write(mo_data)
    
    print(f"Successfully compiled '{po_file}' to '{mo_file}'")
    print(f"Total translations: {num_entries}")

if __name__ == '__main__':
    if len(sys.argv) < 2:
        print("Usage: python3 compile-mo.py <input.po> [output.mo]")
        sys.exit(1)
    
    po_file = sys.argv[1]
    mo_file = sys.argv[2] if len(sys.argv) > 2 else None
    
    compile_po_to_mo(po_file, mo_file)

File: languages/compile-mo.php
<?php
/**
 * Simple PO to MO compiler
 */

if ($argc < 2) {
    echo "Usage: php compile-mo.php <input.po> [output.mo]\n";
    exit(1);
}

$inputFile = $argv[1];
$outputFile = $argv[2] ?? str_replace('.po', '.mo', $inputFile);

if (!file_exists($inputFile)) {
    echo "Error: Input file '$inputFile' not found\n";
    exit(1);
}

// Read PO file
$poContent = file_get_contents($inputFile);
$lines = explode("\n", $poContent);

$translations = [];
$currentMsgid = '';
$currentMsgstr = '';
$inMsgid = false;
$inMsgstr = false;

foreach ($lines as $line) {
    $line = trim($line);
    
    if (empty($line) || $line[0] === '#') {
        continue;
    }
    
    if (strpos($line, 'msgid ') === 0) {
        if ($inMsgstr) {
            $translations[$currentMsgid] = $currentMsgstr;
        }
        $currentMsgid = substr($line, 7, -1);
        $currentMsgstr = '';
        $inMsgid = true;
        $inMsgstr = false;
    } elseif (strpos($line, 'msgstr ') === 0) {
        $currentMsgstr = substr($line, 8, -1);
        $inMsgid = false;
        $inMsgstr = true;
    } elseif ($inMsgid && strpos($line, '"') === 0) {
        $currentMsgid .= substr($line, 1, -1);
    } elseif ($inMsgstr && strpos($line, '"') === 0) {
        $currentMsgstr .= substr($line, 1, -1);
    }
}

if ($inMsgstr) {
    $translations[$currentMsgid] = $currentMsgstr;
}

// Create MO file
$moData = '';
$moData .= pack('V', 0x950412de); // Magic number
$moData .= pack('V', 0); // Version
$moData .= pack('V', count($translations)); // Number of entries
$moData .= pack('V', 28); // Offset of table with original strings
$moData .= pack('V', 28 + 8 * count($translations)); // Offset of table with translation strings

$originalStrings = '';
$translationStrings = '';
$originalOffsets = [];
$translationOffsets = [];

foreach ($translations as $original => $translation) {
    $originalOffsets[] = [
        'length' => strlen($original),
        'offset' => strlen($originalStrings)
    ];
    $originalStrings .= $original . "\0";
    
    $translationOffsets[] = [
        'length' => strlen($translation),
        'offset' => strlen($translationStrings)
    ];
    $translationStrings .= $translation . "\0";
}

// Write original strings table
foreach ($originalOffsets as $offset) {
    $moData .= pack('V', $offset['length']);
    $moData .= pack('V', $offset['offset']);
}

// Write translation strings table
foreach ($translationOffsets as $offset) {
    $moData .= pack('V', $offset['length']);
    $moData .= pack('V', $offset['offset']);
}

$moData .= $originalStrings;
$moData .= $translationStrings;

file_put_contents($outputFile, $moData);

echo "Successfully compiled '$inputFile' to '$outputFile'\n";
echo "Total translations: " . count($translations) . "\n";